pkg/provider/provider.go

```diff
@@ -147,67 +128,122 @@ func (p *Provider) buildConfigForService(t *topology.Topology, cfg *dynamic.Conf
 	return nil
 }
 
+func (p *Provider) buildMiddlewaresForConfigFromService(cfg *dynamic.Configuration, svc *topology.Service) ([]string, error) {
+	var middlewareKeys []string
+
+	middlewares, err := p.buildServiceMiddleware(svc.Annotations)
+	if err != nil {
+		return middlewareKeys, fmt.Errorf("unable to build middlewares: %w", err)
+	}
+
+	for name, middleware := range middlewares {
+		middlewareKey := getMiddlewareKey(svc, name)
+		cfg.HTTP.Middlewares[middlewareKey] = middleware
+
+		middlewareKeys = append(middlewareKeys, middlewareKey)
+	}
+
+	return middlewareKeys, nil
+}
+
+func (p *Provider) buildConfigRoutersAndServices(t *topology.Topology, cfg *dynamic.Configuration, svc *topology.Service, scheme, trafficType string, middlewareKeys []string) error {
+	err := p.buildServicesAndRoutersForService(t, cfg, svc, scheme, trafficType, middlewareKeys)
+	if err != nil {
+		return fmt.Errorf("unable to build routers and services: %w", err)
+	}
+
+	return nil
+}
+
+func (p *Provider) buildACLConfigRoutersAndServices(t *topology.Topology, cfg *dynamic.Configuration, svc *topology.Service, scheme, trafficType string, middlewareKeys []string) {
```

Comment:
I'm a bit surprised that this method does not return an error. I think it should also return an error to indicate to the caller that something is wrong. Then, the caller could inspect the topology for further details. WDYT?

Reply:
@jspdown was there a reason that this code didn't return an error and merely logged the error in the topology?

Reply:
Will open a follow up issue for this.

Reply:
https://github.com/containous/maesh/issues/585